doc/volume-permissions-spec.md

```diff
@@ -0,0 +1,67 @@
+### Introduction
+
+Currently the Docker volume driver for VSphere allows any container on any VM with access to a given datastore
+to create, mount, and delete volumes at will. There are no limits to the size of the volumes that
+can be created or quotas for datastore usage for container volumes. This lack of access control
+takes the management of storage out of the hands of the IT administrator and puts it in the hands of
+the developer. This is a worrying proposition to many IT admins, and since they are the ones that
+must allow deployment of the driver and it's associated host agent, we must allow them to regain
+control with proper enforcement mechanisms that limit damage due to malfeasance or mistake.
+
+### Permissions Model
+
+Since docker containers run inside VMs, and can have root permissions, we cannot rely on the VM user
+for access control. Permissions must be granted to a VM, since VMs are created and managed
+by the IT administrator. This brings up a few issues:
+
+  1. How are VMs identified in order to grant permissions?
+  2. What is the granularity of a permission?
+  3. Where are permissions stored?
+  4. What are the actual permissions?
+
+In order to answer these questions, we must consider 2 scenarios. The first scenarios is permissions
+for a single VM with a local datastore, and the second is multiple VMs with a shared datastore. The
+model must work for both scenarios in order to be useful.
+
+1. VMs can share names so must be identified by unique IDs. These IDs should be unique across a
+   cluster of hosts. If desired, we can associate many IDs with a Role that can be used to grant
+   permissions. In this case permissions granted to a role apply to all VMs that have that role
+   applied.
+2. Permissions are managed per datastore. This means that VMs can have different privileges on
+   different datastores. For instance, on one datastore a VM may be allowed to create volumes of a
+   given size, while on another datastore the VM cannot create volumes at all. Note that in order to
+   provide out of the box usability and experimentation, all datastores allow all operations with no
+   restrictions. Permissions must be added per VM per Datastore by an administrator as necessary.
+3. Since there is no useful centralized storage outside of VC for storing permissions, and we don't
```

Comment:
Can we consider classifying storage into categories or classes (based on capabilities defined commonly) and not expose storage as datastores. And allow users a quota or entitlement on that storage. The idea of needing users to "know" what datastores are available at the backend (that too without knowing their capability, we aren't advertising anything about the storage itself) and then deciding where to place their volumes makes it "non-SDDC-like" (:-) ).  Users may crowd on the same datastore giving the system no way to optimize placement.

Instead, the user should be able to query what classes/categories of storage are available in terms of storage capabilities (TBD) and then acquire an entitlement on those (as they need or say choose to pay for). Then the user gets to create volumes against that entitlement. And the system has the flexibility to "decide" where to locate volumes for that user based on their entitlements.

Its possible that multiple datastores available on a host are all the same category, and the system would consolidate like storage into pools or containers and then allocations are made on that pool - across datastores in the pool).

So we have abstractions like,

a. Storage pools
b. Pool elements - datastores (disks)
c. Pool properties  - essentially the capabilities of the pool's member datastores.
d. Pool stats.
e. Pools ops - add/remove datastores (disks).

Thats a more user friendly approach and abstracts the backend completely than exposing backend datastores.